app/components/base/breadcrumbs/breadcrumb.service.ts

```diff
@@ -0,0 +1,145 @@
+import { Injectable } from "@angular/core";
+import { ActivatedRoute, NavigationEnd, PRIMARY_OUTLET, Router } from "@angular/router";
+import { BehaviorSubject, Observable } from "rxjs";
+
+import { Constants } from "app/utils";
+import { Breadcrumb, RouteComponent } from "./breadcrumb-models";
+
+function breadcrumbMethodMessage(componentName) {
+    const message = "The breadcrumb could not be generated because the route component"
+        + ` '${componentName}' doesn't have the static breadcrumb method defined`;
+    return `${message}
+    class ${componentName} {
+        // Add this method
+        public static breadcrumb(params, queryParams) {
+            return {name: "Some name", label: "Some label"};
+        }
+    }
+`;
+}
+
+@Injectable()
+export class BreadcrumbService {
+    public crumbs: Observable<Breadcrumb[]>;
+    public _crumbs = new BehaviorSubject<Breadcrumb[]>([]);
+
+    constructor(private router: Router, private activatedRoute: ActivatedRoute) {
+        this._loadInitialData();
+        this.router.events.filter(event => event instanceof NavigationEnd).subscribe(event => {
+            let root: ActivatedRoute = this.activatedRoute.root;
+            const crumb = this.getBreadcrumbFromRoute(root);
+            if (crumb) {
+                this.addBreadcrumb(crumb);
+            }
+        });
+        this.crumbs = this._crumbs.asObservable();
+        this._crumbs.subscribe((data) => {
+            sessionStorage.setItem(Constants.sessionStorageKey.breadcrumb, JSON.stringify(data));
+        });
+    }
+
+    public addBreadcrumb(breadcrumb: Breadcrumb) {
+        const cleaned = this._cleanupCrumbs(breadcrumb);
+        this._crumbs.next(cleaned.concat([breadcrumb]));
+    }
+
+    /**
+     * Compare segments of 2 breadcrumb
+     * If the 2 breadcrumb don't have the same base return false
+     */
+    public compareSegments(first: Breadcrumb, second: Breadcrumb) {
+        const firstSegments = first.segments;
+        const secondSegments = second.segments;
+
+        for (let i = 0; i < Math.min(firstSegments.length, secondSegments.length); i++) {
```

Comment:
by "same base" what do you mean? Just the same first item, or their entire path as to be the same? Depending on the answer, you may consider an early-exit by checking the lengths of the segments first.